src/mi/Gaddress_validator.c

```diff
@@ -0,0 +1,349 @@
+{
+  pipe (pipefd);
+  _set_pipe_flags(pipefd[0]);
+  _set_pipe_flags(pipefd[1]);
+}
+#endif
+
+
+static void
+_open_pipe (void)
+{
+  if (_mem_validate_pipe[0] != -1)
+    close (_mem_validate_pipe[0]);
+  if (_mem_validate_pipe[1] != -1)
+    close (_mem_validate_pipe[1]);
+
+  _do_pipe2 (_mem_validate_pipe);
+}
+
+
+/**
+ * Test is a memory address is valid by trying to write from it
+ * @param[in]  addr The address to validate
+ *
+ * @returns true of the memory address is valid (readable), false otherwise.
+ *
+ * This check works by using the address as a (one-byte) buffer in a
+ * write-to-pipe operation.  The write will fail if the memory is not in the
+ * process's address space and marked as readable.
+ */
+static bool
+_write_validate (unw_word_t addr)
+{
+  int ret = -1;
+  ssize_t bytes = 0;
+
+  do
+    {
+      char buf;
+      bytes = read (_mem_validate_pipe[0], &buf, 1);
+    }
+  while ( errno == EINTR );
+
+  if (!(bytes > 0 || errno == EAGAIN || errno == EWOULDBLOCK))
+    {
+      // re-open closed pipe
+      _open_pipe ();
+    }
+
+  do
+    {
+#ifdef HAVE_SYS_SYSCALL_H
+      /* use syscall insteadof write() so that ASAN does not complain */
+      ret = syscall (SYS_write, _mem_validate_pipe[1], addr, 1);
+#else
+      ret = write (_mem_validate_pipe[1], (void *)addr, 1);
+#endif
+    }
+  while ( errno == EINTR );
+
+  return ret > 0;
+}
+
+
+static bool
+_msync_validate (unw_word_t addr, size_t len)
+{
+  if (msync ( (void *)unw_page_start (addr), len, MS_ASYNC) != 0)
+    {
+      return false;
+    }
+
+  return _write_validate (addr);
+}
+
+
+#ifdef HAVE_MINCORE
+static bool
+_mincore_validate (unw_word_t addr, size_t len)
+{
+  unsigned char mvec[2]; /* Unaligned access may cross page boundary */
+
+  /* mincore could fail with EAGAIN but we conservatively return false
+     instead of looping. */
+  if (mincore ((void *)unw_page_start (addr), len, mvec) != 0)
+    {
+      return false;
+    }
+
+  return _write_validate (addr);
+}
+#endif
+
+
+static void
+_unw_address_validator_init(void)
+{
+  _open_pipe ();
+
+  /* Work out dynamically what memory validation function to use. */
+#ifdef HAVE_MINCORE
+  unsigned char present = 1;
+  size_t len = unw_page_size;
+  unw_word_t addr = unw_page_start((unw_word_t)&present);
+  unsigned char mvec[1];
+  int ret;
+  do
+    {
+      ret = mincore ((void*)addr, len, mvec);
+    }
+  while (ret == -1 && errno == EAGAIN);
+  if (ret == 0)
+    {
+      Debug(1, "using mincore to validate memory\n");
+      _mem_validate_func = _mincore_validate;
+    }
+  else
+#endif
+    {
+      Debug(1, "using msync to validate memory\n");
+      _mem_validate_func = _msync_validate;
+    }
+  _unw_address_validator_initialized = ~0;
+}
+
+/* Cache of already validated addresses */
+enum { NLGA = 4 };
+
+#if defined(HAVE___CACHE_PER_THREAD) && HAVE___CACHE_PER_THREAD
+// thread-local variant
+static _Thread_local unw_word_t last_good_addr[NLGA];
+static _Thread_local int lga_victim;
```

Comment:
TLS variables may also trigger `malloc` in glibc, so how do we avoid to access this before being in a signal context in every thread?

Reply:
The malloc in libc (GNU libc or otherwise) is when the DSO is loaded or the thread created. Threads do not receive signals before they're created and if a thread is doing a dlopen() directly from a signal context it's already solidly in undefined-behaviour territory. Only function-scoped thread-local storage triggers a memory allocation after load time or thread creation time, and these are file-scoped thread-locals.

Also, this code is copied and pasted from the existing x86_64, aarch64, s390x, and loongarch64 source files as a deduplication so it has presumably been proven in production for quite some time. We have unit tests that verify malloc is not called during unwinding.